### ISSUE TITLE
[Material] Multiline SnackBar

### ISSUE THREAD

**Issue by orestesgaolin** [flutter/flutter#61949]
## Use case

The Material Library specifies a component that can be described as multiline SnackBar, where the action button is placed below the content in the bottom right corner:

![image](https://user-images.githubusercontent.com/16854239/87995179-17972f80-caef-11ea-852c-8d494545aa29.png)

https://material.io/components/snackbars#anatomy

## Proposal

As Flutter user I would love to be able to use this kind of SnackBar that would dynamically adapt to its content.

If there's enough space then show both action and content in one line, but if not then move action button below the content.

I created initial dynamic SnackBar as a proof of concept with usage of `MultiChildRenderObjectWidget`available [here on DartPad](https://dartpad.dartlang.org/2bc1f8a061d884a64b3154ffb3efc7cc). Of course it needs some tweaking around placement and spacing, but I'm using it successfully in one of my apps.

I would love to get some feedback around that and if you find this approach acceptable I can propose a PR extending current SnackBar behavior. This could be solved by adding enum property `layout` (similar to `behavior`) which would default to current behavior and allow to switch to adaptive one.

cc: @willlarche 

![CleanShot 2020-07-21 at 13 29 18](https://user-images.githubusercontent.com/16854239/88049943-3ed51780-cb56-11ea-90b8-c43fba83efe4.gif)

**Comment by willlarche** [comment]
Thank you!

I'm trying to think why we'd want to keep the old behavior. This seems like preferred behavior for both l10n and a11y/text scaling.

**Comment by orestesgaolin** [comment]
Let me know @willlarche if you'd like to see more mature version of this that complies to the specs. 

The specification images show that there are two separate layouts that user can choose from. Thus, I guess it should be up to the user if they want to stick to the default behavior or switch to the _adaptive_ similarly to how the `floating` flag works.

My proposal would be to add some parameter similar to `behavior`, e.g. `layout` that would default to `SnackBarLayout.actionInline` with possibility to change to `SnackBarLayout.actionAdaptive`.

I'm not the best in naming things so feel free to propose something else :D

![image](https://user-images.githubusercontent.com/16854239/88070062-8b2e5080-cb72-11ea-8bde-78d6fbdeb3fd.png)

  https://material.io/components/snackbars#specs

**Comment by justinmc** [comment]
This proposal looks great, thanks for this.  I just wanted to comment on the technical approach and say that the code looks good at a high level.  Using MultiChildRenderObjectWidget is the first thing that came to my mind for this.

And I'm also confused by that example in the spec above!  The "longer action text" one.  I would think that the message wouldn't wrap.

**Comment by orestesgaolin** [comment]
I've just discussed this case with one of fellow designers and they have a bit different understanding than me as a developer.

They would decide to use multi-line layout when the _action button_ is longer than usual, not when the _content_ is longer (so opposite assumption to mine). Thus, in their opinion, placing of the button below the content is an explicit action, not a result of adaptive layout. What do you think?

**Comment by willlarche** [comment]
Also @rami-a do you know about this at all?

**Comment by willlarche** [comment]
We can get some comment from the Material Designers on how they meant for
this to work.

**Comment by rami-a** [comment]
I agree that this proposal seems reasonable, I'll follow up with our designers to get some input on this and report back as to what the intentions are. Thanks @orestesgaolin

**Comment by rami-a** [comment]
Hey @orestesgaolin I confirmed with our designers that this proposal is reasonable. We should keep the default behavior of being inline for backwards compatibility but providing an alternative (through an enum like you mentioned) makes sense.

Your demo makes sense to me, though please ensure that the action keeps the height it had in the existing layout to ensure adequate tap target size. It looks like your demo shrinks the height a bit.

**Comment by orestesgaolin** [comment]
Yes, I used a bit smaller height in the demo but then reverted to default hit test size in my project. 

Thanks for clarifying! I can propose a PR with tests for backwards compatibility in the near future. Is the approach with `MultiChildRenderObjectWidget` acceptable in your opinion? 

If you take a look at the gist you'll notice that I need to set the size at the beginning of `performLayout` because later I need this to layout `action` with `BoxConstraints.loose(size)`:


```dart
    final width = constraints.maxWidth;
    if (constraints.hasBoundedHeight) {
      final height = constraints.maxHeight;
      size = constraints.constrain(Size(width, height));
    } else {
      size = constraints.constrain(Size(width, 10000.0)); //this could be any other number
    }
```

This seems to be a bit wrong. How do you feel about it?

**Comment by rami-a** [comment]
I think using `MultiChildRenderObjectWidget` makes sense. I'm not sure I quite follow this question though, are you concerned with using the 10000? Perhaps this should be double.infinity to indicate that the height is allowed to grow as much as it needs to?
